File: csharp/Ixgbe/Agent.cs
using System;
using System.Collections.Generic;
using System.Threading;
using TinyNF.Environment;
using TinyNF.Unsafe;

namespace TinyNF.Ixgbe
{
    public ref struct Agent
    {
        private readonly Array256<PacketData> _packets;
        private readonly Array256<Descriptor> _receiveRing;
        private readonly Array256Array<Descriptor> _transmitRings;
        private readonly Ref<uint> _receiveTail;
        private readonly Span<TransmitHead> _transmitHeads;
        private readonly RefArray<uint> _transmitTails;
        private readonly Array256<ushort> _outputs; // trade off a tiny bit of unused space for no bounds checks
        private byte _processDelimiter;


        public Agent(IEnvironment env, Device inputDevice, IReadOnlyList<Device> outputDevices)
        {
            _processDelimiter = 0;
            _outputs = new Array256<ushort>(s => env.Allocate<ushort>(s).Span);

            _packets = new Array256<PacketData>(s => env.Allocate<PacketData>(s).Span);

            _transmitRings = new Array256Array<Descriptor>(outputDevices.Count, s => env.Allocate<Descriptor>(s).Span);
            for (int r = 0; r < _transmitRings.Length; r++)
            {
                for (int n = 0; n < 256; n++)
                {
                    _transmitRings[r][(byte)n].Buffer = Endianness.ToLittle(env.GetPhysicalAddress(ref _packets[(byte)n]));
                }
            }

            _receiveRing = _transmitRings[0];
            _receiveTail = new Ref<uint>(inputDevice.SetInput(env, _receiveRing.AsSpan()).Span);

            _transmitHeads = env.Allocate<TransmitHead>((uint)outputDevices.Count).Span;
            _transmitTails = new RefArray<uint>(outputDevices.Count);
            for (byte n = 0; n < outputDevices.Count; n++)
            {
                _transmitTails[n] = outputDevices[n].AddOutput(env, _transmitRings[n].AsSpan(), ref _transmitHeads[n].Value).Span;
            }
        }

        public void Run(PacketProcessor processor)
        {
            nint n;
            for (n = 0; n < DriverConstants.FlushPeriod; n++)
            {
                ulong receiveMetadata = Endianness.FromLittle(Volatile.Read(ref _receiveRing[_processDelimiter].Metadata));
                if ((receiveMetadata & (1ul << 32)) == 0)
                {
                    break;
                }

                ushort length = (ushort)receiveMetadata;
                processor(ref _packets[_processDelimiter], length, _outputs);

                ulong rsBit = ((_processDelimiter % DriverConstants.RecyclePeriod) == (DriverConstants.RecyclePeriod - 1)) ? (1ul << (24 + 3)) : 0ul;

                // not clear why we have to copy _transmitRings here (its only member is an array), but this is necessary for the bounds check to be eliminated
                // using a byte as the index likewise doesn't lead to the bounds check being eliminated
                var _transmitRings = this._transmitRings;
                for (int b = 0; b < _transmitRings.Length; b++)
                {
                    _transmitRings[b][_processDelimiter].Metadata = _outputs[(byte)b] | rsBit | (1ul << (24 + 1)) | (1ul << 24);
                    _outputs[(byte)b] = 0;
                }

                _processDelimiter++;

                if (rsBit != 0)
                {
                    uint earliestTransmitHead = _processDelimiter;
                    ulong minDiff = 0xFF_FF_FF_FF_FF_FF_FF_FFul;
                    foreach (ref var headRef in _transmitHeads)
                    {
                        uint head = Endianness.FromLittle(Volatile.Read(ref headRef.Value));
                        ulong diff = head - _processDelimiter;
                        if (diff <= minDiff)
                        {
                            earliestTransmitHead = head;
                            minDiff = diff;
                        }
                    }

                    Volatile.Write(ref _receiveTail.Get(), Endianness.ToLittle((earliestTransmitHead - 1) % DriverConstants.RingSize));
                }
            }
            if (n != 0)
            {
                foreach (ref var tail in _transmitTails)
                {
                    Volatile.Write(ref tail, Endianness.ToLittle(_processDelimiter));
                }
            }
        }
    }
}

File: rust/src/ixgbe/agent.rs
use crate::env::environment::Environment;
use crate::lifed::{LifedArray, LifedPtr, LifedSlice};

use super::device::{Descriptor, Device, PacketData, TransmitHead, RING_SIZE, RX_METADATA_DD, RX_METADATA_LENGTH, TX_METADATA_EOP, TX_METADATA_IFCS, TX_METADATA_LENGTH, TX_METADATA_RS};

pub const MAX_OUTPUTS: usize = 256; // so that an u8 can index into it without bounds checks

const FLUSH_PERIOD: u8 = 8;
const RECYCLE_PERIOD: u8 = 64;

// OVERHEAD: each of these slices has the same length, but we can't state it in code
pub struct Agent<'a> {
    buffers: &'a mut [PacketData; RING_SIZE],
    rings: LifedSlice<'a, LifedArray<'a, Descriptor, { RING_SIZE }>>,
    receive_tail_addr: LifedPtr<'a, u32>,
    transmit_heads: LifedSlice<'a, TransmitHead>,
    transmit_tail_addrs: LifedSlice<'a, LifedPtr<'a, u32>>,
    outputs: &'a mut [u64; MAX_OUTPUTS],
    processed_delimiter: u8,
}

impl<'a> Agent<'a> {
    pub fn create(env: &impl Environment<'a>, input: LifedPtr<'a, Device<'a>>, outputs: &[LifedPtr<'a, Device<'a>>]) -> Agent<'a> {
        // LifedSlice requires a nonzero length, allowing us to access the shared first ring without any checks
        if outputs.len() == 0 {
            panic!("Cannot have zero outputs");
        }
        if outputs.len() >= MAX_OUTPUTS {
            panic!("Too many outputs");
        }

        let buffers = env.allocate::<PacketData, { RING_SIZE }>();

        let rings = env.allocate_slice::<LifedArray<'a, Descriptor, { RING_SIZE }>>(outputs.len());
        for r in 0..outputs.len() {
            rings[r] = LifedArray::new(env.allocate::<Descriptor, { RING_SIZE }>());
        }
        for n in 0..buffers.len() {
            let packet_phys_addr = u64::to_le(env.get_physical_address(&mut buffers[n as usize]) as u64);
            for r in 0..outputs.len() {
                rings[r].index(n as usize).write_volatile_part(packet_phys_addr, |d| &mut d.addr);
            }
        }

        let receive_tail_addr = input.map(|d| d.set_input(env, rings[0].index(0)));

        let transmit_heads = LifedSlice::new(env.allocate_slice::<TransmitHead>(outputs.len()));
        let transmit_tail_addrs = LifedSlice::new(env.allocate_slice::<LifedPtr<'a, u32>>(outputs.len()));
        for n in 0..outputs.len() {
            transmit_tail_addrs.set(n, outputs[n].map(|d| d.add_output(env, rings[n].index(0), transmit_heads.index(n))));
        }

        Agent {
            buffers,
            rings: LifedSlice::new(rings),
            receive_tail_addr,
            transmit_heads,
            transmit_tail_addrs,
            outputs: env.allocate::<u64, { MAX_OUTPUTS }>(),
            processed_delimiter: 0,
        }
    }

    #[inline(always)] // mimic C "static inline"
    pub fn run(&mut self, processor: fn(&mut PacketData, u64, &mut [u64; MAX_OUTPUTS])) {
        let mut n: u8 = 0;
        while n < FLUSH_PERIOD {
            let receive_metadata = u64::from_le(self.rings.first().index(self.processed_delimiter as usize).read_volatile_part(|d| &d.metadata));
            if (receive_metadata & RX_METADATA_DD) == 0 {
                break;
            }

            let length = RX_METADATA_LENGTH(receive_metadata);
            processor(&mut self.buffers[self.processed_delimiter as usize], length, self.outputs);

            let rs_bit: u64 = if self.processed_delimiter % RECYCLE_PERIOD == (RECYCLE_PERIOD - 1) { TX_METADATA_RS } else { 0 };
            for o in 0..self.rings.len() {
                self.rings
                    .get(o as usize)
                    .index(self.processed_delimiter as usize)
                    .write_volatile_part(u64::to_le(TX_METADATA_LENGTH(self.outputs[o as u8 as usize]) | rs_bit | TX_METADATA_IFCS | TX_METADATA_EOP), |d| {
                        &mut d.metadata
                    });
                self.outputs[o as u8 as usize] = 0;
            }

            self.processed_delimiter = self.processed_delimiter.wrapping_add(1); // modulo ring size implicit since it's an u8

            if rs_bit != 0 {
                let mut earliest_transmit_head = self.processed_delimiter as u32;
                let mut min_diff = u64::MAX;
                for h in 0..self.transmit_heads.len() {
                    let head_value = u32::from_le(self.transmit_heads.index(h).read_volatile_part(|h| &h.value));
                    let diff = (head_value as u64).wrapping_sub(self.processed_delimiter as u64);
                    if diff <= min_diff {
                        earliest_transmit_head = head_value;
                        min_diff = diff;
                    }
                }

                self.receive_tail_addr.write_volatile(u32::to_le(earliest_transmit_head));
            }
            n += 1;
        }
        if n != 0 {
            for t in 0..self.transmit_tail_addrs.len() {
                self.transmit_tail_addrs.get(t).write_volatile(u32::to_le(self.processed_delimiter as u32));
            }
        }
    }
}
